crates/astria-sequencer-client/src/client.rs

```diff
@@ -17,95 +18,41 @@ use eyre::{
 };
 use tendermint::block::Height;
 use tendermint_rpc::{
-    endpoint::{
-        block::Response as BlockResponse,
-        broadcast::{
-            tx_commit,
-            tx_sync,
-        },
-        validators,
+    endpoint::broadcast::{
+        tx_commit,
+        tx_sync,
     },
     Client as _,
-    HttpClient,
 };
 
 /// Default Tendermint base URL.
 pub const DEFAULT_TENDERMINT_BASE_URL: &str = "http://localhost:26657";
 
 /// Tendermint HTTP client which is used to interact with the Sequencer node.
-pub struct Client {
-    client: HttpClient,
+#[async_trait]
+pub trait Client {
+    async fn get_balance(&self, address: &Address, height: Option<Height>)
+    -> eyre::Result<Balance>;
+    async fn get_nonce(&self, address: &Address, height: Option<Height>) -> eyre::Result<Nonce>;
+    async fn submit_transaction_sync(&self, tx: Signed) -> eyre::Result<tx_sync::Response>;
+    async fn submit_transaction_commit(&self, tx: Signed) -> eyre::Result<tx_commit::Response>;
 }
 
-impl Client {
-    /// Creates a new Tendermint client with the given base URL.
-    ///
-    /// # Errors
-    ///
-    /// Returns an error if the client could not be created.
-    pub fn new(base_url: &str) -> eyre::Result<Self> {
-        Ok(Client {
-            client: HttpClient::new(base_url).wrap_err("failed to initialize tendermint client")?,
-        })
-    }
-
-    /// Returns the latest block.
-    ///
-    /// # Errors
-    ///
-    /// - If calling the tendermint RPC endpoint fails.
-    pub async fn get_latest_block(&self) -> eyre::Result<BlockResponse> {
-        let block = self
-            .client
-            .latest_block()
-            .await
-            .wrap_err("failed to call latest_block")?;
-        Ok(block)
-    }
-
-    /// Returns the block at the given height.
-    ///
-    /// # Errors
-    ///
-    /// - If calling the tendermint RPC endpoint fails.
-    pub async fn get_block(&self, height: Height) -> eyre::Result<BlockResponse> {
-        let block = self
-            .client
-            .block(height)
-            .await
-            .wrap_err("failed to call block")?;
-        Ok(block)
-    }
-
-    /// Returns the validator set at the given height.
-    ///
-    /// # Errors
-    ///
-    /// - If calling the tendermint RPC endpoint fails.
-    pub async fn get_validator_set<T: Into<Height>>(
-        &self,
-        height: T,
-    ) -> eyre::Result<validators::Response> {
-        self.client
-            .validators(height.into(), tendermint_rpc::Paging::Default)
-            .await
-            .wrap_err("failed to get validator set")
-    }
-
+#[async_trait]
+impl Client for tendermint_rpc::HttpClient {
```

Comment:
You can avoid all these explicit impls by making `ClientExt: Client` a supertrait of `tendermint_rpc::Client` and then providing default trait methods. That way all of these functions will immediately work for both, the http and the websocket clients.